Source/WebKit/WebProcess/WebPage/WebFrame.cpp

```diff
@@ -1510,4 +1510,25 @@ void WebFrame::setAppBadge(const WebCore::SecurityOriginData& origin, std::optio
     send(Messages::WebFrameProxy::SetAppBadge(origin, badge));
 }
 
+std::optional<ResourceResponse> WebFrame::resourceResponseForURL(const URL& url) const {
+    RefPtr localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
+    if (!localFrame)
+        return std::nullopt;
+
+    RefPtr loader = localFrame->loader().documentLoader();
+    if (!loader)
+        return std::nullopt;
+
+    // First, try the main resource.
+    if (loader->url() == url)
+        return loader->response();
+
+    // Next, try subresources.
+    RefPtr<ArchiveResource> resource = loader->subresource(url);
```

Comment:
I don't think these comments add any additional understanding to anyone reading the code.  Let's omit them.
Also, deduction guides can make it so you don't need to explicitly state the type as a template parameter, and the scope of this can be in the if statement:
if (RefPtr resource = loader->subresource(url))